File: MATLAB/write_windstress_bb86.m
%% write windstress double gyre

   clear all,clf, close all
   iodir='/Net/yucatan/abozec/BB86_PACKAGE/MATLAB/';
   addpath(genpath(['',iodir,'/UTILITIES/']));   
   
   %% PATH
   io = [iodir,'/../topo/'];
   file_grid = 'regional.grid.BB86.a'

   %% domain
   idm = 101 ;
   jdm = 101 ;
   tdm = 12 ;%% monthly files

   %% name the new files
   file_E = 'forcing.tauewd.BB86';  %% without .a or .b %%
   file_N = 'forcing.taunwd.BB86';


   %% Read grid
   [plon,plat]=read_grid_hycom(idm, jdm, ([io,'../topo/']), file_grid);

   %% Calculation of the analytical wind-stress (N/m2)
   ustress = zeros(jdm);
   stressa = -1.; 
   sconv = 1.e-1;   %% scale factor form dyn/cm2 to N/m2

   %% BB86 formulation
   for j = 1:jdm
     ustress(j) = stressa*cos(double(j-1)/double(jdm-1)*6.28318530718)*sconv;
   end

   %% plot the stress 
   figure(1)
   plot(ustress,plat);
   ylim([0 20]);
   xlim([-0.2 0.2]);
   grid on;
   

   %% Write the wind-stress files
   tte = zeros(jdm, idm, tdm);
   ttn = zeros(jdm, idm, tdm);
  
   IJDM=idm*jdm;
   npad=4096-mod(IJDM,4096);
   toto=zeros(npad,1);

 
   %% we apply  a taux , no tauy
   for j = 1:jdm 
     tte(j, :, :) = ustress(j);
   end  

   %% Taux
   taux_fid=fopen([io,'../force/',file_E,'.a'],'w');   

   for ll = 1:tdm 
     A=tte(:,:,ll)';
     A=reshape(A,IJDM,1);
     %% Writing the field
     fwrite(taux_fid,A,'float32','ieee-be');
     %% Writing the padding at the end of the record:
     fwrite(taux_fid,toto,'float32','ieee-be');
   end 
   fclose(taux_fid);
   
   %% Tauy
   tauy_fid=fopen([io,'../force/',file_N,'.a'],'w');   

   for ll = 1:tdm 
     A=ttn(:,:,ll)';
     A=reshape(A,IJDM,1);
     %% Writing the field
     fwrite(tauy_fid,A,'float32','ieee-be');
     %% Writing the padding at the end of the record:
     fwrite(tauy_fid,toto,'float32','ieee-be');
   end 
   fclose(tauy_fid);
   

   %% create .b file 
   time = 1:12;
   
   %% Write Eastward Wind stress file
   fldb=[io,'../force/',file_E,'.b'];
   fid1=fopen(fldb,'wt');

   fprintf(fid1, 'Analytical Eastward Wind-stress\n');
   fprintf(fid1, '\n');
   fprintf(fid1, '\n');
   fprintf(fid1, '\n');
   fprintf(fid1, 'i/jdm =  %i  %i \n',idm,jdm);
   for m = 1:tdm 
      fprintf(fid1,' tauewd: month,range =  %2.2i  %10.5E  %10.5E\n', ...
		   time(m), min(min(tte(:, :, m)')), max(max(tte(:, :, m)')));
   end 
        
   fclose(fid1);

   %% Write Northward Wind stress file
   fldb=[io,'../force/',file_N,'.b'];
   fid1=fopen(fldb,'wt');

   fprintf(fid1, 'Analytical Northward Wind-stress\n');
   fprintf(fid1, '\n');
   fprintf(fid1, '\n');
   fprintf(fid1, '\n');
   fprintf(fid1, 'i/jdm =  %i  %i \n',idm,jdm);
   for m = 1:tdm 
      fprintf(fid1,' taunwd: month,range =  %2.2i  %10.5E  %10.5E\n', ...
		   time(m), min(min(ttn(:, :, m)')), max(max(ttn(:, :, m)')));
   end 
        
   fclose(fid1);

  

File: PYTHON/write_windstress_bb86.py
#!python3

# create relax files
import sys
import os
from os.path import join
### might not need next line in older or newer version of proplot
os.environ['PROJ_LIB'] = '/Users/abozec/opt/anaconda3/share/proj'
import proplot as plot
import numpy as np

iodir='/Users/abozec/Documents/GitHub/BB86_PACKAGE/PYTHON/'
sys.path.append(iodir)
from hycom.io import read_hycom_grid

io=iodir+'../topo/'
pl = 0 ##  1 or 0 for plot or not

## size of the domain
idm = 101 ; jdm = 101
tdm = 12 ## 12 month climatology

## name the new files
file_E = 'forcing.tauewd.BB86.python'
file_N = 'forcing.taunwd.BB86.python'

## Read grid
grid_data=join(io,'regional.grid.BB86.python.a')
#print(F"The fields available are: {read_field_grid_names(grid_data)}")
fieldg=['plon','plat']
grid_field= read_hycom_grid(grid_data, fieldg)
plon=grid_field['plon'][:,:]
plat=grid_field['plat'][:,:]

## Calculation of the analytical wind-stress (N/m2)
ustress = np.zeros([jdm])
stressa = -1.
sconv = 1.e-1   ## scale factor form dyn/cm2 to N/m2

## BB86 formulation
for j in np.arange(jdm):
   ustress[j] = stressa*np.cos(int(j-1)/int(jdm-1)*6.28318530718)*sconv

## figure
fig, ax = plot.subplots()
ax.plot(ustress, plat[:,0], lw=2)
ax.format(suptitle='Wind-stress', xlim=(-0.2,0.2),ylim=(0,20))
plot.show(block=False)

## Write the wind-stress files
tte = np.zeros([jdm, idm, tdm])
ttn = np.zeros([jdm, idm, tdm])

## get size of pad for writing the file
layer_size=idm*jdm
npad = 4096-np.mod(layer_size, 4096)

## we apply  a taux , no tauy
for j in np.arange(jdm):
   tte[j, :, :] = ustress[j]

## open .a file eastward wind
fout=open(iodir+'../force/'+file_E+'.a','wb')

## write .a file
for t in np.arange(tdm):
   field32=np.array(tte[:,:,t],dtype='>f4') ## float32 big_endian
   ## writing .a file
   fout.write(np.reshape(field32,layer_size))
   if (npad != 4096):
      fout.write(np.zeros([npad],dtype='>f4'))

fout.close()

# writing .b file
fbout=open(iodir+'../force/'+file_E+'.b','w')
fbout.write("Analytical Eastward Wind-stress \n")
fbout.write("  \n")
fbout.write("  \n")
fbout.write("  \n")
fbout.write(F"i/jdm = {idm},{jdm}  \n")
for t in np.arange(tdm):
   fbout.write("tauewd:  month,range = {0:02d} {1:10.5E} {2:10.5E} \n".\
             format(t+1,np.nanmin(tte[:,:,t]),np.nanmax(tte[:,:,t])))

fbout.close()


## open .a file northward wind
fout=open(iodir+'../force/'+file_N+'.a','wb')

## write .a file
for t in np.arange(tdm):
   field32=np.array(ttn[:,:,t],dtype='>f4') ## float32 big_endian
   ## writing .a file
   fout.write(np.reshape(field32,layer_size))
   if (npad != 4096):
      fout.write(np.zeros([npad],dtype='>f4'))

fout.close()

# writing .b file
fbout=open(iodir+'../force/'+file_N+'.b','w')
fbout.write("Analytical Northward Wind-stress \n")
fbout.write("  \n")
fbout.write("  \n")
fbout.write("  \n")
fbout.write(F"i/jdm = {idm},{jdm}  \n")
for t in np.arange(tdm):
   fbout.write("taunwd:  month,range = {0:02d} {1:10.5E} {2:10.5E} \n".\
             format(t+1,np.nanmin(ttn[:,:,t]),np.nanmax(ttn[:,:,t])))

fbout.close()
